taichi/codegen/codegen_llvm.h

```diff
@@ -1428,8 +1424,8 @@ class CodeGenLLVM : public IRVisitor, public ModuleBuilder {
 
       current_coordinates = new_coordinates;
 
-      // Additional compare if non-POT exists
-      auto nonpot_cond = tlctx->get_constant(true);
+      // Additional compare if non-POT exists or leaf block is bitmasked
+      auto additional_cond = tlctx->get_constant(true);
```

Comment:
See if I'm understanding correctly -- By `&& bitmasked_is_active`, this implicitly becomes a `1x1x1` mask, even if the actual dimension of this last `bitmasked` layer could be larger (e.g. `4x8x8`?).

BTW, since Taichi is not limited to support 3D, this mask can be `1x1x...x1` (eight of `1`) in theory..?

Reply:
Oh, I feel like `additional_cond` is a bit confusing because it's unclear to me what `additional` is. Since it determines whether we want to run the kernel at this particular coordinate or not, maybe `execute_cond` (or `run_cond`, etc)? Naming is always hard :/

Reply:
> See if I'm understanding correctly -- By `&& bitmasked_is_active`, this implicitly becomes a `1x1x1` mask, even if the actual dimension of this last `bitmasked` layer could be larger (e.g. `4x8x8`?).

Right!



> BTW, since Taichi is not limited to support 3D, this mask can be `1x1x...x1` (eight of `1`) in theory..?

Exactly, we can do up to 8-D. I'm just using `1x1x1` since our world is 3D.



> Oh, I feel like `additional_cond` is a bit confusing because it's unclear to me what `additional` is. Since it determines whether we want to run the kernel at this particular coordinate or not, maybe `execute_cond` (or `run_cond`, etc)? Naming is always hard :/

Good point. Changed to `exec_cond` and added more comments.

Reply:
Thanks! One more thing i just realized: do we need to do this for `pointer` as well (and any other sparse snodes)?

Reply:
Good question! My quick answer is no, since using (64bit + one memory indirection) for a pointer addressing at the leaf level is too expensive compared to `bitmasked`. 

Here we are only generating for the leaf level (small blocks such as `4x4x4`), and the only suitable data structure here for sparsity is a bitmask...

Reply:
Pointers are still useful for representing the sparsity of larger blocks, such as whether a `4x4x4` block is active, since its cost is amortized over the 64 voxels. It is faster than hash table, and uses less memory (since it's dynamically allocated) compared to bitmask.